File: leetcode_2594/MinimumTimeToRepairCars.java
package leetcode_2594;

import java.util.Arrays;

class MinimumTimeToRepairCars {
    public long repairCars(int[] ranks, int cars) {
        long left = 1;
        long right = (long) Arrays.stream(ranks).min().getAsInt() * cars * cars;
        long res = -1;

        while(left <= right)
        {
            long mid = left + (right - left) / 2;
            long count = countRepaired(mid, ranks);
            if(count >= cars)
            {
                res = mid;
                right = mid - 1;
            }
            else
                left = mid + 1;
        }
        return res;
    }

    private long countRepaired(long time, int[] ranks)
    {
        long count = 0;
        for(int rank : ranks)
            count += Math.sqrt(time / rank);
        return count;
    }

    public static void main(String[] args) {
        MinimumTimeToRepairCars obj = new MinimumTimeToRepairCars();
        int[] ranks = { 4, 2, 3, 1 };
        int cars = 10;
        System.out.println(obj.repairCars(ranks, cars));
    }
}

File: leetcode_2594/MinimumTimeToRepairCars.py
from math import sqrt


class MinimumTimeToRepairCars:
    def repairCars(self, ranks: list[int], cars: int) -> int:
        
        def count_repaired(time):
            count = 0
            for rank in ranks:
                count += int(sqrt(time / rank))
            return count
        
        left = 1
        right = min(ranks) * cars * cars
        res = -1
        
        while left <= right:
            mid = left + (right - left) // 2
            count = count_repaired(mid)
            if count >= cars:
                res = mid
                right = mid - 1
            else:
                left = mid + 1
        return res
    
def main():
    ranks = [4, 3, 2, 1]
    cars = 10
    result = MinimumTimeToRepairCars().repairCars(ranks, cars)
    print(result)
    
if __name__ == "__main__":
    main()